wire/msgmixpairreq.go

```diff
@@ -0,0 +1,426 @@
+// Copyright (c) 2023 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package wire
+
+import (
+	"bytes"
+	"fmt"
+	"hash"
+	"io"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+const (
+	// MaxMixPairReqScriptClassLen is the maximum length allowable for a
+	// MsgMixPairReq script class.
+	MaxMixPairReqScriptClassLen = 32
+
+	// MaxMixPairReqUTXOs is the maximum number of unspent transaction
+	// outputs that may be contributed in a single mixpairreq message.
+	MaxMixPairReqUTXOs = 512 // XXX: PNOOMA
+
+	// MaxMixPairReqUTXOScriptLen is the maximum length allowed for the
+	// unhashed P2SH script of a UTXO ownership proof.
+	// XXX: might want to limit this to standard script sizes
+	MaxMixPairReqUTXOScriptLen = 16384 // txscript.MaxScriptSize
+
+	// MaxMixPairReqUTXOPubKeyLen is the maximum length allowed for the
+	// pubkey of a UTXO ownership proof.
+	MaxMixPairReqUTXOPubKeyLen = 33
+
+	// MaxMixPairReqUTXOSignatureLen is the maximum length allowed for the
+	// signature of a UTXO ownership proof.
+	MaxMixPairReqUTXOSignatureLen = 64
+)
+
+// MixPairReqUTXO describes an unspent transaction output to be spent in a
+// mix.  It includes a proof that the output is able to be spent, by
+// containing a signature and the necessary data needed to prove key
+// possession.
+type MixPairReqUTXO struct {
+	OutPoint  OutPoint
+	Script    []byte // Only used for P2SH
+	PubKey    []byte
+	Signature []byte
+}
+
+// MsgMixPairReq implements the Message interface and represents a mixing pair
+// request message.  It describes a type of coinjoin to be created, unmixed
+// data being contributed to the coinjoin, and proof of ability to sign the
+// resulting coinjoin.
+type MsgMixPairReq struct {
+	Signature    [64]byte
+	Identity     [33]byte
+	Expiry       uint32
+	MixAmount    int64
+	ScriptClass  string
+	TxVersion    uint16
+	LockTime     uint32
+	MessageCount uint32
+	InputValue   int64
+	UTXOs        []MixPairReqUTXO
+	Change       *TxOut
+
+	// hash records the hash of the message.  It is a member of the
+	// message for convenience and performance, but is never automatically
+	// set during creation or deserialization.
+	hash chainhash.Hash
+}
+
+// Pairing returns a description of the coinjoin transaction being created.
+// Different mixpairreq messages area compatible to perform a mix together if
+// their pairing descriptions are identical.
+func (msg *MsgMixPairReq) Pairing() ([]byte, error) {
+	w := bytes.NewBuffer(make([]byte, 0, 8+32+2+4))
+
+	err := writeElement(w, msg.MixAmount)
+	if err != nil {
+		return nil, err
+	}
+
+	err = WriteVarString(w, MixVersion, msg.ScriptClass)
```

Comment:
Looks like there is no check on the length of the script class string here, but it's required to be `MaxMixPairReqScriptClassLen`.

Reply:
Added this check at the top of this function:


```
	// Require script class to be strict ASCII and not exceed the maximum
	// length.
	lenScriptClass := len(msg.ScriptClass)
	if lenScriptClass > MaxMixPairReqScriptClassLen {
		msg := fmt.Sprintf("script class length is too long "+
			"[len %d, max %d]", lenScriptClass,
			MaxMixPairReqScriptClassLen)
		return messageError(op, ErrMixPairReqScriptClassTooLong, msg)
	}
	if !isStrictAscii(msg.ScriptClass) {
		msg := "script class string is not strict ASCII"
		return messageError(op, ErrMalformedStrictString, msg)
	}
```